### ISSUE TITLE
Question about params: yolov8x vs yolov8x-p2  

### ISSUE THREAD

**Issue by soyeollee** [ultralytics/ultralytics#3607]
### Search before asking

- [x] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello. 

I'm trying to use the yolov8x-p2 model.

Num params of YOLOv8x =  68229648, but YOLOv8x-p2 = 66637536

In detection layer, YOLOv8x-p2 has fewer parameters.

Why is there such a difference?

Thank you!

* I used [yolov8x.yaml, yolov8x-p2.yaml] without modification.
----

## YOLOv8x-p2 

```
                   from  n    params  module                                       arguments                     
  0                  -1  1      2320  ultralytics.nn.modules.conv.Conv             [3, 80, 3, 2]                 
  1                  -1  1    115520  ultralytics.nn.modules.conv.Conv             [80, 160, 3, 2]               
  2                  -1  3    436800  ultralytics.nn.modules.block.C2f             [160, 160, 3, True]           
  3                  -1  1    461440  ultralytics.nn.modules.conv.Conv             [160, 320, 3, 2]              
  4                  -1  6   3281920  ultralytics.nn.modules.block.C2f             [320, 320, 6, True]           
  5                  -1  1   1844480  ultralytics.nn.modules.conv.Conv             [320, 640, 3, 2]              
  6                  -1  6  13117440  ultralytics.nn.modules.block.C2f             [640, 640, 6, True]           
  7                  -1  1   3687680  ultralytics.nn.modules.conv.Conv             [640, 640, 3, 2]              
  8                  -1  3   6969600  ultralytics.nn.modules.block.C2f             [640, 640, 3, True]           
  9                  -1  1   1025920  ultralytics.nn.modules.block.SPPF            [640, 640, 5]                 
 10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 11             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 12                  -1  3   7379200  ultralytics.nn.modules.block.C2f             [1280, 640, 3]                
 13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 14             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 15                  -1  3   1948800  ultralytics.nn.modules.block.C2f             [960, 320, 3]                 
 16                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 17             [-1, 2]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 18                  -1  3    488000  ultralytics.nn.modules.block.C2f             [480, 160, 3]                 
 19                  -1  1    230720  ultralytics.nn.modules.conv.Conv             [160, 160, 3, 2]              
 20            [-1, 15]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 21                  -1  3   1795200  ultralytics.nn.modules.block.C2f             [480, 320, 3]                 
 22                  -1  1    922240  ultralytics.nn.modules.conv.Conv             [320, 320, 3, 2]              
 23            [-1, 12]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 24                  -1  3   7174400  ultralytics.nn.modules.block.C2f             [960, 640, 3]                 
 25                  -1  1   3687680  ultralytics.nn.modules.conv.Conv             [640, 640, 3, 2]              
 26             [-1, 9]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 27                  -1  3   7379200  ultralytics.nn.modules.block.C2f             [1280, 640, 3]                
 28    [18, 21, 24, 27]  1   4688976  ultralytics.nn.modules.head.Detect           [80, [160, 320, 640, 640]]    
YOLOv8x-p2 summary: 445 layers, 66637536 parameters, 66637520 gradients
```
----

## YOLOv8x 
```
                   from  n    params  module                                       arguments                     
  0                  -1  1      2320  ultralytics.nn.modules.conv.Conv             [3, 80, 3, 2]                 
  1                  -1  1    115520  ultralytics.nn.modules.conv.Conv             [80, 160, 3, 2]               
  2                  -1  3    436800  ultralytics.nn.modules.block.C2f             [160, 160, 3, True]           
  3                  -1  1    461440  ultralytics.nn.modules.conv.Conv             [160, 320, 3, 2]              
  4                  -1  6   3281920  ultralytics.nn.modules.block.C2f             [320, 320, 6, True]           
  5                  -1  1   1844480  ultralytics.nn.modules.conv.Conv             [320, 640, 3, 2]              
  6                  -1  6  13117440  ultralytics.nn.modules.block.C2f             [640, 640, 6, True]           
  7                  -1  1   3687680  ultralytics.nn.modules.conv.Conv             [640, 640, 3, 2]              
  8                  -1  3   6969600  ultralytics.nn.modules.block.C2f             [640, 640, 3, True]           
  9                  -1  1   1025920  ultralytics.nn.modules.block.SPPF            [640, 640, 5]                 
 10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 11             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 12                  -1  3   7379200  ultralytics.nn.modules.block.C2f             [1280, 640, 3]                
 13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 14             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 15                  -1  3   1948800  ultralytics.nn.modules.block.C2f             [960, 320, 3]                 
 16                  -1  1    922240  ultralytics.nn.modules.conv.Conv             [320, 320, 3, 2]              
 17            [-1, 12]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 18                  -1  3   7174400  ultralytics.nn.modules.block.C2f             [960, 640, 3]                 
 19                  -1  1   3687680  ultralytics.nn.modules.conv.Conv             [640, 640, 3, 2]              
 20             [-1, 9]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 21                  -1  3   7379200  ultralytics.nn.modules.block.C2f             [1280, 640, 3]                
 22        [15, 18, 21]  1   8795008  ultralytics.nn.modules.head.Detect           [80, [320, 640, 640]]         
YOLOv8x summary: 365 layers, 68229648 parameters, 68229632 gradients
```

**Comment by glenn-jocher** [comment]
@soyeollee hi there,

Thank you for your question. The difference in the number of parameters between the YOLOv8x-p2 and YOLOv8x models is due to the architectural differences in the detection layers.

The YOLOv8x-p2 model has fewer parameters in the detection layer because it has a reduced number of convolutional filters for detecting objects. This reduction in parameters helps optimize the model's memory usage and computational efficiency, making it more suitable for certain applications.

In contrast, the YOLOv8x model has a higher number of parameters in the detection layer as it uses a larger number of convolutional filters, which can potentially provide better accuracy and detection performance in certain scenarios.

It's worth noting that the specific choices for the number of parameters and architectural design are often based on trade-offs between accuracy, computational requirements, and efficiency.

I hope this explanation clarifies the difference in parameter counts between YOLOv8x-p2 and YOLOv8x for you. If you have any further questions, feel free to ask!

Best regards.

**Comment by soyeollee** [comment]
@glenn-jocher Thank you for you answer :)

In yolov5x-p2,  In since the number of channels (v5x, v5x-p2) is the same, it increases by the parameter of p2.

So, I wondered if the reduction of the parameters was intended.

In my case(dataset), num of channelsof original v8x-p2 head are too small.

So, I modified `head.py/class Detect/def __init__` as follows:

```
    def __init__(self, nc=80, ch=()):  # detection layer
        super().__init__()
        self.nc = nc  # number of classes
        self.nl = len(ch)  # number of detection layers
        self.reg_max = 16  # DFL channels (ch[0] // 16 to scale 4/8/12/16/20 for n/s/m/l/x)
        self.no = nc + self.reg_max * 4  # number of outputs per anchor
        self.stride = torch.zeros(self.nl)  # strides computed during build

        if len(ch) == 4:  # add p2 layer
            # CHANGE
            c2, c3 = max((16, ch[1] // 4, self.reg_max * 4)), max(ch[1], min(self.nc, 100))  # [160 320 640 1280]   
            # c2, c3 = max((16, ch[0] // 4, self.reg_max * 4)), max(ch[0], min(self.nc, 100))  # [320 640 1280] 
        else:
            # ORIGIANL
            c2, c3 = max((16, ch[0] // 4, self.reg_max * 4)), max(ch[0], min(self.nc, 100))  # [320 640 1280]
```

![Screenshot from 2023-07-10 09-47-20](https://github.com/ultralytics/ultralytics/assets/59550240/c2aa3ac2-ffde-4ee6-9a70-ace41cdb7021)


Have a good day :)
